urllib3/util/connection.py

```diff
@@ -108,6 +116,134 @@ def _set_socket_options(sock, options):
         sock.setsockopt(*opt)
 
 
+def _connect_happy_eyes(address, timeout=socket._GLOBAL_DEFAULT_TIMEOUT,
+                        source_address=None, socket_options=None):
+    """ Implements the Happy Eyes protocol (RFC 6555) which allows
+    multiple sockets to attempt to connect from different families
+    for better connect times for dual-stack clients where server
+    IPv6 service is advertised but broken. """
+    err = None
+
+    if address in HAPPY_EYES_CACHE:
+        family, proto, expires = HAPPY_EYES_CACHE[address]
```

Comment:
Well, technically the dictionary is thread-safe. The problem here is the LBYL logic in place here. Ideally we should just refactor this to a `try...except`.